dns/client.go

```diff
@@ -0,0 +1,114 @@
+
+	"github.com/miekg/dns"
+)
+
+// Resolver is the interface that wraps the Resolve method.
+//
+// Resolve resolves a domain name to an IP address. It returns a slice of IP
+// addresses as strings.
+type Resolver interface {
+	Resolve(ctx context.Context, query, recordType string, nameserver Nameserver) ([]string, error)
+}
+
+// Lookuper is the interface that wraps the Lookup method.
+//
+// As opposed to a Resolver which uses a specific nameserver to resolve the
+// query, a Lookuper uses the system's default resolver.
+//
+// Lookup resolves a domain name to an IP address. It returns a slice of IP
+// addresses as strings.
+type Lookuper interface {
+	Lookup(ctx context.Context, hostname string) ([]string, error)
+}
+
+// Client is a DNS resolver that uses the `miekg/dns` package under the hood.
+//
+// It implements the Resolver interface.
+type Client struct {
+	// client is the DNS client used to resolve queries.
+	client dns.Client
+}
+
+// Ensure our Client implements the Resolver interface
+var _ Resolver = &Client{}
+
+// Ensure our Client implements the Lookuper interface
+var _ Lookuper = &Client{}
+
+// NewDNSClient creates a new Client.
+func NewDNSClient() *Client {
+	return &Client{
+		client: dns.Client{},
+	}
+}
+
+// Resolve resolves a domain name to a slice of IP addresses using the given nameserver.
+// It returns a slice of IP addresses as strings.
+func (r *Client) Resolve(
+	ctx context.Context,
+	query, recordType string,
+	nameserver Nameserver,
+) ([]string, error) {
+	concreteType, err := RecordTypeString(recordType)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"resolve operation failed with %w, %s is an invalid DNS record type",
+			ErrUnsupportedRecordType,
+			recordType,
+		)
+	}
+
+	// Prepare the DNS query message
+	//
+	// Because the dns package [dns.SetQuestion] function expects specific
+	// uint16 values for the record type, and we don't want to leak that
+	// to our public API, we need to convert our RecordType to the
+	// corresponding uint16 value.
+	message := dns.Msg{}
+	message.SetQuestion(query+".", uint16(concreteType))
+
+	// Query the nameserver
+	response, _, err := r.client.ExchangeContext(ctx, &message, nameserver.Addr())
+	if err != nil {
+		return nil, fmt.Errorf("querying the dns nameserver failed: %w", err)
+	}
+
+	if response.Rcode != dns.RcodeSuccess {
+		return nil, newDNSError(response.Rcode, "DNS query failed")
+	}
+
+	var ips []string
+	for _, a := range response.Answer {
+		switch t := a.(type) {
+		case *dns.A:
+			ips = append(ips, t.A.String())
+		case *dns.AAAA:
+			ips = append(ips, t.AAAA.String())
+		default:
+			return nil, fmt.Errorf(
+				"resolve operation failed with %w: unhandled DNS answer type %T",
+				ErrUnsupportedRecordType,
+				a,
+			)
+		}
+	}
+
+	return ips, nil
+}
+
+// Lookup resolves a domain name to a slice of IP addresses using the system's
+// default resolver.
+func (r *Client) Lookup(ctx context.Context, hostname string) ([]string, error) {
```

Comment:
Is Lookup needed? Couldn't the same problem be addressed by using a default value for resolve?

Reply:
I've asked myself the same question. My personal take on this in the context of k6 is "I prefer to have it", mostly for API consistency reasons.

On the one hand, I think the distinction between resolution (asking a specific DNS server for specific kinds of records) and lookup (asking the OS to talk with their configured DNS servers and get A records for a given host) is valuable.

On the other hand, we could embed (call under the hood) the lookup operation as part of the resolve operation when neither the `recordType`, nor the `nameserver` arguments are provided. But that would lead to a "conditional" API with an implicit behavior: if you give me a single argument, I'll do X; if you give me three of them, I'll do Y, and instinctively I find those APIs error-prone.

But that's really me, I don't hold a strong opinion, and I'm happy to discuss alternatives 🙇🏻

Reply:
I don't have a strong opinion, but as you mentioned you got inspired by Deno, I see they are doing a unique method instead of two: https://docs.deno.com/examples/dns-queries.

Reply:
I had completely missed this example, and had based my inspiration on the `resolveDNS` API docs which doesn't seem to mention it performs a lookup under the hood.